File: day1/rooks/solutions/kostka.py
import collections


def calculate_moves(A, K):
    N = len(A)
    M = len(A[0])
    total_elements = N * M

    pos = [None] * (total_elements + 1)
    for r in range(N):
        for c in range(M):
            pos[A[r][c]] = (r, c)

    INF = float('inf')
    dist = [INF] * (total_elements + 1)

    dist[1] = 0

    row_deques = [collections.deque() for _ in range(N)]
    col_deques = [collections.deque() for _ in range(M)]

    if total_elements > 0:
        start_r, start_c = pos[1]
        row_deques[start_r].append((0, 1))
        col_deques[start_c].append((0, 1))

    for val in range(2, total_elements + 1):
        r, c = pos[val]

        min_moves = INF

        while row_deques[r] and row_deques[r][0][1] < val - K:
            row_deques[r].popleft()
        if row_deques[r]:
            min_moves = min(min_moves, row_deques[r][0][0] + 1)

        while col_deques[c] and col_deques[c][0][1] < val - K:
            col_deques[c].popleft()
        if col_deques[c]:
            min_moves = min(min_moves, col_deques[c][0][0] + 1)

        dist[val] = min_moves

        while row_deques[r] and row_deques[r][-1][0] >= dist[val]:
            row_deques[r].pop()
        row_deques[r].append((dist[val], val))

        while col_deques[c] and col_deques[c][-1][0] >= dist[val]:
            col_deques[c].pop()
        col_deques[c].append((dist[val], val))

    result_matrix = [[0] * M for _ in range(N)]
    for i in range(1, total_elements + 1):
        r, c = pos[i]
        result_matrix[r][c] = -1 if dist[i] == INF else int(dist[i])

    return result_matrix

File: day1/rooks/solutions/kostka.java
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Deque;

public class rooks {

    private static class DistValue {
        int dist;
        int val;

        DistValue(int dist, int val) {
            this.dist = dist;
            this.val = val;
        }
    }

    public static int[][] calculate_moves(int[][] A, int K) {
        int N = A.length;
        int M = A[0].length;
        int totalElements = N * M;

        int[][] pos = new int[totalElements + 1][2];
        for (int r = 0; r < N; r++) {
            for (int c = 0; c < M; c++) {
                pos[A[r][c]][0] = r;
                pos[A[r][c]][1] = c;
            }
        }

        final int INF = Integer.MAX_VALUE;
        int[] dist = new int[totalElements + 1];
        Arrays.fill(dist, INF);

        if (totalElements > 0) {
            dist[1] = 0;
        }

        Deque<DistValue>[] rowDeques = new ArrayDeque[N];
        Deque<DistValue>[] colDeques = new ArrayDeque[M];
        for (int i = 0; i < N; i++) {
            rowDeques[i] = new ArrayDeque<>();
        }
        for (int i = 0; i < M; i++) {
            colDeques[i] = new ArrayDeque<>();
        }

        if (totalElements > 0) {
            int startR = pos[1][0];
            int startC = pos[1][1];
            rowDeques[startR].addLast(new DistValue(0, 1));
            colDeques[startC].addLast(new DistValue(0, 1));
        }

        for (int val = 2; val <= totalElements; val++) {
            int r = pos[val][0];
            int c = pos[val][1];

            int minMoves = INF;

            Deque<DistValue> rowDeque = rowDeques[r];
            while (!rowDeque.isEmpty() && rowDeque.peekFirst().val < val - K) {
                rowDeque.pollFirst();
            }
            if (!rowDeque.isEmpty()) {
                int predDist = rowDeque.peekFirst().dist;
                if (predDist != INF) {
                    minMoves = Math.min(minMoves, predDist + 1);
                }
            }

            Deque<DistValue> colDeque = colDeques[c];
            while (!colDeque.isEmpty() && colDeque.peekFirst().val < val - K) {
                colDeque.pollFirst();
            }
            if (!colDeque.isEmpty()) {
                int predDist = colDeque.peekFirst().dist;
                if (predDist != INF) {
                    minMoves = Math.min(minMoves, predDist + 1);
                }
            }
            
            dist[val] = minMoves;

            if (dist[val] != INF) {
                while (!rowDeque.isEmpty() && rowDeque.peekLast().dist >= dist[val]) {
                    rowDeque.pollLast();
                }
                rowDeque.addLast(new DistValue(dist[val], val));

                while (!colDeque.isEmpty() && colDeque.peekLast().dist >= dist[val]) {
                    colDeque.pollLast();
                }
                colDeque.addLast(new DistValue(dist[val], val));
            }
        }

        int[][] resultMatrix = new int[N][M];
        for (int i = 1; i <= totalElements; i++) {
            int r = pos[i][0];
            int c = pos[i][1];
            resultMatrix[r][c] = (dist[i] == INF) ? -1 : dist[i];
        }

        return resultMatrix;
    }
}